feat(insights): aurelia insights plugin

Vheissu opened the pull request:
This pull request adds an entirely new package `insights` that is standalone and installable for developers to measure the performance of their Aurelia 2 applications. It features automatic lifecycle measurement, to allow developers to measurement performance and identify potential bottlenecks (slow lifecycle methods, long-running asynchronous API requests, etc).

This relates to issue https://github.com/aurelia/aurelia/issues/2166 after brief discussion in the Aurelia team discord channel around the development of this plugin.

Vheissu committed:
feat(insights): scaffolding initial plugin

Sets up the basic structure and configuration files for the insights package.
This includes adding:
- `package.json` with dependencies and build scripts
- `rollup.config.mjs` for bundling
- `tsconfig.json` for TypeScript compilation
- `.eslintrc.cjs` and `.eslintignore` for linting
- `CHANGELOG.md` for tracking changes

This is based on similar structure to the other packages

Vheissu committed:
feat(insights): add performance insights

- Interfaces for performance measurements and configurations
- A performance tracker service to start, end, and collect measurements
- Lifecycle hooks to automatically measure component lifecycles
- Router lifecycle hooks to track routing performance
- Utilities for grouping, analysing, and exporting performance data

Vheissu committed:
feat(insights): configuration

Introduces a configuration system for insights, allowing customisation of tracking options and filters. Oh, it also provides a builder pattern for convenient configuration.

Vheissu committed:
feat(insights): telemetry, typings and plugin root exports/logic

- Core interfaces and types for insights configuration.
- Performance tracking via a `PerformanceTracker`.
- Telemetry capabilities using `TelemetryMeter` and `ActivitySource`.

Vheissu committed:
chore(insights): configure monorepo

Adds the `@aurelia/insights` package to the monorepo.

Vheissu committed:
refactor(insights): interfaces to use singletons

Updates `IPerformanceTracker`, `ITelemetryService`, and
`IInsightsUtilities` interfaces to extend their
corresponding implementation classes and register the implementations as singletons.

Vheissu committed:
feat(insights): configuration tests

Adds tests for the insights configuration to ensure correct default values, option merging, and builder functionality.

Vheissu committed:
feat(insights): utilities tests

This commit introduces comprehensive unit tests for the Insights Utilities class.

These tests cover various functionalities, including measurement retrieval, clearing, grouping by component, and calculation of statistics for slowest and most active components.

Vheissu committed:
refactor(insights): simplify dependency registration

Refactors the insights plugin to use direct container registration for core services. We create default implementations in the classes themselves, not using Registration methods.

Vheissu committed:
chore(insights): fix up paths

Turns out we needed to add our plugin to lots of other places too

Vheissu committed:
fix(insights): needed router packages for types

Made them dev and optional dependencies

Vheissu committed:
feat(insights): v1 implements repeat performance tracking

Adds performance tracking for the `repeat` template controller.

This change introduces a `PerformanceRepeat` class that extends the original `Repeat` to provide detailed performance measurements for rendering and updating collections. It tracks binding, attaching, detaching, unbinding, and items/collection changes.

This could be flaky and might be removed due to possible changes in the root repeater attribute, but will stand for now.

Vheissu committed:
chore(insights): insights package readme

Adds details about the insights package

Vheissu committed:
test(insights): telemetry service tests

These tests cover meter and activity source management, integration scenarios, cleanup, disposal, edge cases, error handling, and dependency injection to ensure the service functions correctly under various conditions.

Vheissu committed:
test(insights): telemetry meter unit tests

tests for the TelemetryMeter, Counter, Histogram, and Gauge classes.

Vheissu committed:
chore(insights): explicit exports

Proper named exports instead of generic ones

Vheissu committed:
chore(insights): forgotten export of performance repeat

probably not needed, but just in case

Vheissu committed:
test(insights): testing the performance repeat

Adds comprehensive unit tests for the PerformanceRepeat class to ensure accurate tracking of lifecycle events and
data changes. These tests cover binding, attaching, detaching, unbinding, itemsChanged, and handleCollectionChange methods, and include scenarios for both enabled and disabled tracking configurations.

Vheissu committed:
refactor(insights)): explicit plugin exports

We need to see what we're working with

Vheissu committed:
chore(insights): oops

Vheissu committed:
feat(insights): adds telemetry dashboard components

Adds components for visualising telemetry data: a dashboard, metric gauges, and metric charts.

Vheissu committed:
feat(insights): event listener binding performance tracking

Adds a patch to `ListenerBinding` to track the performance of event listener calls.

This enables insights into the time spent executing event listeners, providing valuable data for performance analysis and optimisation.

Includes tests to verify the patch and its tracking functionality.

Vheissu committed:
feat(insights): changes to performance metric debug

Vheissu committed:
test(insights): enhance performance repeat tests

Refactors and expands performance repeat tests for improved
coverage and reliability, ensuring accurate performance tracking.

This includes mocking parent class methods to simplify tests,
adding checks for different change types in collection updates,
and enhancing setup and teardown for consistent test execution.

Vheissu committed:
refactor(insights): updates insights parsing logic

Refactors insights parsing to align with changes in the data structure.

Now correctly parses nested component data (grouped, slowest, mostActive) and eventHandlers.

Vheissu committed:
docs(insights): insights package documentation

Adds comprehensive documentation for the `@aurelia/insights` package.

This documentation covers installation, configuration, automatic framework tracking, custom telemetry APIs, visualization components, performance analysis utilities, real-world examples, API reference, best practices, browser support, and troubleshooting.

bigopon reviewed (commented):
Nice work @Vheissu , before we go to the code review, can we make sure the PR has some visual examples for what the end results like? Also if we are borrowing code or ideas from somewhere, can we put some acknowledgement in the files that have the borrowed code/ideas?

Vheissu commented:
> Nice work @Vheissu , before we go to the code review, can we make sure the PR has some visual examples for what the end results like? Also if we are borrowing code or ideas from somewhere, can we put some acknowledgement in the files that have the borrowed code/ideas?

Absolutely. I'm pretty wrecked today, but tomorrow plan on doing those final flourishes. Around attribution (especially the Telemetry API stuff), need to do a sweep and clean of the codebase (I hacked a few things in some parts I need to revisit in the interest of getting things working), cleaning tests and the other little final pieces for completion.

Sayan751 commented:
Can we also integrate this plugin to our e2e tests to detect any performance regression? Or even arrange something to identify the hot areas and/or optimization opportunities in the core packages?

Vheissu committed:
feat(insights): adds performance testing support

Introduces a comprehensive performance testing framework.

Test mode support, allowing collection of performance data without relying on
browser performance APIs.

Adds helpers for generating test data, checking for regressions, and asserting performance thresholds.

Enables configuration of tracking mode to support devtools-only, test-mode, or both. Includes test mode configuration options such as enabling thresholds, collecting detailed metadata, etc.

Vheissu committed:
Merge branch 'master' into feat/insights-plugin

Vheissu committed:
Merge branch 'master' into feat/insights-plugin

Vheissu committed:
refactor(insights): refactors insights for new router packages

Updates the insights package to support both the renamed `router-direct` and `router` packages.

This includes:
- Replacing `router-lite` references with `router-direct`.
- Adding a new test suite to verify router lifecycle hooks functionality.
- Modifying the `RouterLifecycleHooks` class to handle different router package signatures and properly extract route information from both packages.

Vheissu committed:
chore(insights): my eslint didn't catch this

Vheissu committed:
Merge branch 'master' into feat/insights-plugin

bigopon committed:
Merge branch 'master' of https://github.com/aurelia/aurelia into feat/insights-plugin

bigopon committed:
fix: update deps, resolve deepscan issues

bigopon committed:
fix: tweak doc

bigopon committed:
fix: some minor cleanup

bigopon reviewed (approved):
nice work @Vheissu , I don't see any issues since this is the first iteration. Though I'd say the nature of this plugin suggests that it should be on a separate repo, rather together with the rest.
